File: chess/pieces/rook.rb
module Chess::Pieces

  ##
  #
  #  Rook
  #  @author Adrien Thébault <me@adrien-thebault.fr>
  #

  module Rook

    def self.possible_moves(game,pos)

      color, ennemy = game[:chessboard][pos[0]][pos[1]]&0b1000, Chess::Game.ennemy(game[:chessboard][pos[0]][pos[1]]&0b1000)
      offset, res, finished_top, finished_bottom, finished_left, finished_right = 0, [], false, false, false, false

      loop do

        unless finished_top

          res.push([pos[0]+offset, pos[1]]) if game[:chessboard][pos[0]+offset][pos[1]].nil? || game[:chessboard][pos[0]+offset][pos[1]]&0b1000 == ennemy
          finished_top = true if (!game[:chessboard][pos[0]+offset][pos[1]].nil? && offset > 0) || pos[0]+offset == 7

        end

        unless finished_bottom

          res.push([pos[0]-offset, pos[1]]) if game[:chessboard][pos[0]-offset][pos[1]].nil? || game[:chessboard][pos[0]-offset][pos[1]]&0b1000 == ennemy
          finished_bottom = true if (!game[:chessboard][pos[0]-offset][pos[1]].nil? && offset > 0) || pos[0]-offset == 0

        end

        unless finished_left

          res.push([pos[0], pos[1]-offset]) if game[:chessboard][pos[0]][pos[1]-offset].nil? || game[:chessboard][pos[0]][pos[1]-offset]&0b1000 == ennemy
          finished_left = true if (!game[:chessboard][pos[0]][pos[1]-offset].nil? && offset > 0) || pos[1]-offset == 0

        end

        unless finished_right

          res.push([pos[0], pos[1]+offset]) if game[:chessboard][pos[0]][pos[1]+offset].nil? || game[:chessboard][pos[0]][pos[1]+offset]&0b1000 == ennemy
          finished_right = true if (!game[:chessboard][pos[0]][pos[1]+offset].nil? && offset > 0) || pos[1]+offset == 7

        end

        break if finished_top && finished_bottom && finished_left && finished_right
        offset += 1

      end

      res

    end

  end

end

File: lib/game/rook/rook.c
/**
*
*  Rook-specific functions
*  @author Adrien Thébault <me@adrien-thebault.fr>
*
*/

/** Dependencies */

#include <stdbool.h>
#include "../game.h"

/**
*
*  Computes every possible move for a Rook at pos and puts result in possible_moves
*
*/

void ChessGame_Rook_PossibleMoves(game *g, unsigned char pos[2], unsigned char possible_moves[POSSIBLE_MOVES_SIZE][2], unsigned char *possible_moves_length) {

  unsigned char color = g->chessboard[pos[0]][pos[1]] & MASK_PLAYER;
  unsigned char ennemy = ChessGame_Ennemy(color), offset = 0, line = 0, col = 0;

  bool finished_top = false, finished_bottom = false, finished_left = false, finished_right = false;
  while(!(finished_top && finished_bottom && finished_left && finished_right)) {

    if(!finished_top) {

      line = pos[0]+offset; col = pos[1];
      if(g->chessboard[line][col] == -1 || (g->chessboard[line][col] & MASK_PLAYER) == ennemy) {

        possible_moves[*possible_moves_length][0] = line;
        possible_moves[*possible_moves_length][1] = col;
        (*possible_moves_length)++;

      }

      if((g->chessboard[line][col] != -1 && offset > 0) || line == 7) finished_top = true;

    }

    if(!finished_bottom) {

      line = pos[0]-offset; col = pos[1];
      if(g->chessboard[line][col] == -1 || (g->chessboard[line][col] & MASK_PLAYER) == ennemy) {

        possible_moves[*possible_moves_length][0] = line;
        possible_moves[*possible_moves_length][1] = col;
        (*possible_moves_length)++;

      }

      if((g->chessboard[line][col] != -1 && offset > 0) || line == 7) finished_bottom = true;

    }

    if(!finished_left) {

      line = pos[0]; col = pos[1]-offset;
      if(g->chessboard[line][col] == -1 || (g->chessboard[line][col] & MASK_PLAYER) == ennemy) {

        possible_moves[*possible_moves_length][0] = line;
        possible_moves[*possible_moves_length][1] = col;
        (*possible_moves_length)++;

      }

      if((g->chessboard[line][col] != -1 && offset > 0) || line == 7) finished_left = true;

    }

    if(!finished_right) {

      line = pos[0]; col = pos[1]+offset;
      if(g->chessboard[line][col] == -1 || (g->chessboard[line][col] & MASK_PLAYER) == ennemy) {

        possible_moves[*possible_moves_length][0] = line;
        possible_moves[*possible_moves_length][1] = col;
        (*possible_moves_length)++;

      }

      if((g->chessboard[line][col] != -1 && offset > 0) || line == 7) finished_right = true;

    }

    offset++;

  }

}
